File: packages/server/app/neuropsych/fluency.py
"""
Category and Letter fluency scoring with clustering and switching analysis.
"""

import logging
from difflib import SequenceMatcher
from typing import Any, Dict, List, Optional

from .stimuli import ANIMAL_SUBCATEGORIES

logger = logging.getLogger(__name__)

_CATEGORY_MEMBERS: Dict[str, set] = {}


def _get_category_set(category: str) -> set:
    """Build a flat set of known valid members for a category."""
    if category not in _CATEGORY_MEMBERS:
        if category == "animals":
            members = set()
            for subcat_words in ANIMAL_SUBCATEGORIES.values():
                members.update(w.lower() for w in subcat_words)
            _CATEGORY_MEMBERS[category] = members
        else:
            _CATEGORY_MEMBERS[category] = set()
    return _CATEGORY_MEMBERS[category]


def _normalize_word(w: str) -> str:
    """Lowercase, strip punctuation, and basic singular/plural normalization."""
    w = w.lower().strip(".,!?;:'\"")
    if w.endswith("es") and len(w) > 4:
        return w[:-2]
    if w.endswith("s") and len(w) > 3:
        return w[:-1]
    return w


def _is_perseveration(word: str, seen: List[str], threshold: float = 0.75) -> bool:
    """Check if word is a fuzzy duplicate of any previously seen word."""
    nw = _normalize_word(word)
    for s in seen:
        ns = _normalize_word(s)
        if ns == nw:
            return True
        if SequenceMatcher(None, ns, nw).ratio() >= threshold:
            return True
    return False


def _find_subcategory(word: str, category: str) -> Optional[str]:
    """Return the subcategory label for a word, if known."""
    if category != "animals":
        return None
    nw = _normalize_word(word)
    for subcat, members in ANIMAL_SUBCATEGORIES.items():
        if nw in [m.lower() for m in members]:
            return subcat
    return None


def score_category_fluency(
    words: List[Dict[str, Any]],
    category: str,
    duration_sec: float = 60.0,
) -> Dict[str, Any]:
    """Score a category fluency trial from word timestamps.

    Args:
        words: list of {word, start, end, confidence, filler} from STT
        category: e.g. "animals", "vegetables"
        duration_sec: test duration (default 60s)

    Returns dict with total_correct, perseverations, intrusions, clusters,
    n_switches, mean_cluster_size, time_bins.
    """
    valid_set = _get_category_set(category)
    content_words = [w for w in words if not w.get("filler", False)]

    correct: List[Dict[str, Any]] = []
    perseverations: List[Dict[str, Any]] = []
    intrusions: List[Dict[str, Any]] = []
    seen_correct: List[str] = []

    for w in content_words:
        word_text = w["word"]
        nw = _normalize_word(word_text)

        if _is_perseveration(word_text, seen_correct):
            perseverations.append({"word": word_text, "time": w.get("start", 0)})
            continue

        if valid_set and nw not in valid_set:
            intrusions.append({"word": word_text, "time": w.get("start", 0)})
            continue

        seen_correct.append(word_text)
        subcat = _find_subcategory(word_text, category)
        correct.append({
            "word": word_text,
            "time": w.get("start", 0),
            "subcategory": subcat,
        })

    # Clustering and switching analysis
    clusters: List[Dict[str, Any]] = []
    current_cluster: List[str] = []
    current_subcat: Optional[str] = None
    n_switches = 0

    for item in correct:
        sc = item.get("subcategory")
        if sc is None:
            if current_cluster:
                clusters.append({"subcategory": current_subcat, "words": list(current_cluster), "size": len(current_cluster)})
            current_cluster = [item["word"]]
            current_subcat = "unknown"
            n_switches += 1
        elif sc != current_subcat:
            if current_cluster:
                clusters.append({"subcategory": current_subcat, "words": list(current_cluster), "size": len(current_cluster)})
            current_cluster = [item["word"]]
            current_subcat = sc
            n_switches += 1
        else:
            current_cluster.append(item["word"])

    if current_cluster:
        clusters.append({"subcategory": current_subcat, "words": list(current_cluster), "size": len(current_cluster)})

    cluster_sizes = [c["size"] for c in clusters]
    mean_cluster_size = sum(cluster_sizes) / max(1, len(cluster_sizes))

    # Time-bin analysis (quartiles)
    bin_size = duration_sec / 4.0
    time_bins = [0, 0, 0, 0]
    for item in correct:
        t = item["time"]
        idx = min(3, int(t / bin_size))
        time_bins[idx] += 1

    return {
        "total_correct": len(correct),
        "perseverations": perseverations,
        "perseveration_count": len(perseverations),
        "intrusions": intrusions,
        "intrusion_count": len(intrusions),
        "clusters": clusters,
        "n_switches": max(0, n_switches - 1),
        "mean_cluster_size": round(mean_cluster_size, 2),
        "time_bins": time_bins,
        "correct_words": correct,
    }


def score_letter_fluency(
    words: List[Dict[str, Any]],
    letter: str,
    duration_sec: float = 60.0,
) -> Dict[str, Any]:
    """Score a letter fluency trial (F, A, S, L, etc.)."""
    content_words = [w for w in words if not w.get("filler", False)]
    letter_lower = letter.lower()

    correct: List[Dict[str, Any]] = []
    perseverations: List[Dict[str, Any]] = []
    intrusions: List[Dict[str, Any]] = []
    seen_correct: List[str] = []

    for w in content_words:
        word_text = w["word"]
        nw = _normalize_word(word_text)

        if _is_perseveration(word_text, seen_correct):
            perseverations.append({"word": word_text, "time": w.get("start", 0)})
            continue

        if not nw.startswith(letter_lower):
            intrusions.append({"word": word_text, "time": w.get("start", 0)})
            continue

        seen_correct.append(word_text)
        correct.append({"word": word_text, "time": w.get("start", 0)})

    bin_size = duration_sec / 4.0
    time_bins = [0, 0, 0, 0]
    for item in correct:
        idx = min(3, int(item["time"] / bin_size))
        time_bins[idx] += 1

    return {
        "total_correct": len(correct),
        "perseverations": perseverations,
        "perseveration_count": len(perseverations),
        "intrusions": intrusions,
        "intrusion_count": len(intrusions),
        "time_bins": time_bins,
        "correct_words": correct,
    }

File: examples/neuro-psych-tests/lib/scoring/fluency.ts
import type { WordTimestamp } from "@whissle/live-assist-core";
import { ANIMAL_SUBCATEGORIES, VEGETABLE_LIST } from "../stimuli/fluency-categories";

function normalize(w: string): string {
  let n = w.toLowerCase().replace(/[.,!?;:'"]/g, "");
  if (n.endsWith("es") && n.length > 4) n = n.slice(0, -2);
  else if (n.endsWith("s") && n.length > 3) n = n.slice(0, -1);
  return n;
}

function isDuplicate(word: string, seen: string[]): boolean {
  const nw = normalize(word);
  return seen.some((s) => normalize(s) === nw);
}

function findSubcategory(word: string): string | null {
  const nw = normalize(word);
  for (const [cat, members] of Object.entries(ANIMAL_SUBCATEGORIES)) {
    if (members.some((m) => m.toLowerCase() === nw)) return cat;
  }
  return null;
}

export interface FluencyResult {
  total_correct: number;
  perseverations: { word: string; time: number }[];
  intrusions: { word: string; time: number }[];
  clusters: { subcategory: string | null; words: string[]; size: number }[];
  n_switches: number;
  mean_cluster_size: number;
  time_bins: [number, number, number, number];
  correct_words: { word: string; time: number; subcategory: string | null }[];
}

export function scoreCategoryFluency(
  words: WordTimestamp[],
  category: "animals" | "vegetables",
  durationSec = 60,
): FluencyResult {
  const validSet = category === "vegetables"
    ? new Set(VEGETABLE_LIST.map((v) => normalize(v)))
    : null;

  const content = words.filter((w) => !w.filler);
  const correct: FluencyResult["correct_words"] = [];
  const perseverations: FluencyResult["perseverations"] = [];
  const intrusions: FluencyResult["intrusions"] = [];
  const seenCorrect: string[] = [];

  for (const w of content) {
    const nw = normalize(w.word);
    if (isDuplicate(w.word, seenCorrect)) {
      perseverations.push({ word: w.word, time: w.start });
      continue;
    }
    if (category === "animals") {
      const sc = findSubcategory(w.word);
      if (!sc) { intrusions.push({ word: w.word, time: w.start }); continue; }
      seenCorrect.push(w.word);
      correct.push({ word: w.word, time: w.start, subcategory: sc });
    } else {
      if (validSet && !validSet.has(nw)) { intrusions.push({ word: w.word, time: w.start }); continue; }
      seenCorrect.push(w.word);
      correct.push({ word: w.word, time: w.start, subcategory: null });
    }
  }

  // Clustering
  const clusters: FluencyResult["clusters"] = [];
  let curCluster: string[] = [];
  let curSub: string | null = null;
  let switches = 0;

  for (const item of correct) {
    if (item.subcategory !== curSub) {
      if (curCluster.length) clusters.push({ subcategory: curSub, words: [...curCluster], size: curCluster.length });
      curCluster = [item.word];
      curSub = item.subcategory;
      switches++;
    } else {
      curCluster.push(item.word);
    }
  }
  if (curCluster.length) clusters.push({ subcategory: curSub, words: [...curCluster], size: curCluster.length });

  const sizes = clusters.map((c) => c.size);
  const meanCluster = sizes.length ? sizes.reduce((a, b) => a + b, 0) / sizes.length : 0;

  const binSize = durationSec / 4;
  const bins: [number, number, number, number] = [0, 0, 0, 0];
  for (const item of correct) {
    const idx = Math.min(3, Math.floor(item.time / binSize));
    bins[idx]++;
  }

  return {
    total_correct: correct.length,
    perseverations,
    intrusions,
    clusters,
    n_switches: Math.max(0, switches - 1),
    mean_cluster_size: Math.round(meanCluster * 100) / 100,
    time_bins: bins,
    correct_words: correct,
  };
}

export function scoreLetterFluency(
  words: WordTimestamp[],
  letter: string,
  durationSec = 60,
): Omit<FluencyResult, "clusters" | "n_switches" | "mean_cluster_size"> {
  const l = letter.toLowerCase();
  const content = words.filter((w) => !w.filler);
  const correct: { word: string; time: number; subcategory: null }[] = [];
  const perseverations: { word: string; time: number }[] = [];
  const intrusions: { word: string; time: number }[] = [];
  const seen: string[] = [];

  for (const w of content) {
    if (isDuplicate(w.word, seen)) { perseverations.push({ word: w.word, time: w.start }); continue; }
    if (!normalize(w.word).startsWith(l)) { intrusions.push({ word: w.word, time: w.start }); continue; }
    seen.push(w.word);
    correct.push({ word: w.word, time: w.start, subcategory: null });
  }

  const binSize = durationSec / 4;
  const bins: [number, number, number, number] = [0, 0, 0, 0];
  for (const item of correct) {
    bins[Math.min(3, Math.floor(item.time / binSize))]++;
  }

  return { total_correct: correct.length, perseverations, intrusions, time_bins: bins, correct_words: correct };
}
